[CORE-1620] archival: Unify segment upload and reupload code paths

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

From the [original PR](https://github.com/redpanda-data/redpanda/pull/25731)

This PR changes the segment upload path.
Previously, we used segment_collector to reupload data (compacted reupload or adjacent segment merging). To upload new segments archival_policy was used. So the code is scattered across two different places. This PR adds new collection mode to the segment_collector. This mode allows segment collector to upload new segments. It works similarly to archival_policy, tries to upload whole segments if possible. Does the scanning if needed. The goal here is to replace implementation of the segment_collector with the one that uses log_reader interface.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none
* 
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

Lazin committed:
archival: Add method that extracts stream

from the 'segment_collector' intance

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

Lazin committed:
archival: Add test that checks

...make_puload_candidate_stream method

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

Lazin committed:
archival: Add offset skipping

"make_upload_candidate_stream" can skip offset now. The behavior is
similar to how ntp_archiver works in this case.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

Lazin committed:
archival: Add test case for offset skipping

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

Lazin committed:
archival: New segment_collector mode

Adds new collector mode which can be used to upload new segments. This
makes 'segment_collector' suitable to replace some logic inside the
'archival_policy'. The long term goal here is to make 'segment_reupload'
the only code path which is used to find upload candidates. Then the
'ntp_archiver' should start using method of the segment collector which
returns a stream instead of the set of segments. When its done the
internals of the 'segment_collector' could be updated to use
'log_reader' instead of iterating through the segments.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

Lazin committed:
archival: Swap segment collector implementation

Previously, the search of segments that can be uploaded to the cloud
storage was implemented inside the archival_policy. For non-compacted
segments the archival_policy used its internal implementation but for
compacted reuploads it was using segment_collector.

This commit introduces new segment serach logic that uses
segment_collector so both compacted and non-compacted uploads are going
through the collector now.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

Lazin committed:
archival: Remove dead code

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>
Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64878-->
<!--16c8111e-2263-4fbe-9384-d5660d8abd9b-->
<!--4094c276-4af1-4914-8c5c-5a30d27088e1-->
<!--42b235a0-2f24-48d2-ba7e-f7b39c06481f-->
<!--641230e9-890c-4d1c-85ca-23ad3d0d0183-->
<!--title-->
## Retry command for [Build#64878](https://buildkite.com/redpanda/redpanda/builds/64878)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/rest_catalog_connection_test.py::RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog@{"cloud_storage_type":1}
tests/rptest/tests/internal_topic_protection_test.py::InternalTopicProtectionLargeClusterTest.test_consumer_offset_topic
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/datalake/disk_budget_test.py::DatalakeDiskUsageTest.test_idle_finish@{"cloud_storage_type":1,"concurrent_translations":4,"num_partitions":10}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3a776545-3783-438c-8c7e-f513ace113bc-->
<!--TEST RESULTS-->
<!--6e6c90d8-ba2b-4287-8106-3066bfb76867-->
<!--TEST RESULTS-->
<!--82211e20-aba9-4b5e-8f4b-91cc639a6f7b-->
<!--TEST RESULTS-->
<!--e29e21e6-e2f7-4a6e-bae1-e76a2c6af365-->
<!--TEST RESULTS-->
<!--5ddd66bd-61d9-4bfb-a748-70b42b35892c-->
<!--TEST RESULTS-->
<!--fb243f7e-f7f8-4947-9ded-54ab3128a83b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64878">build#64878</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.disk_budget_test.DatalakeDiskUsageTest.test_idle_finish.num_partitions=10.concurrent_translations=4.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64878#01965f09-4fc0-4338-9a9f-dd1d9f95a2f1 | FAIL | 0/1 |
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64878#01965f1b-7771-40cd-87ae-e96f3dfa1e3c | FLAKY | 20/21 |
| rptest.tests.internal_topic_protection_test.InternalTopicProtectionLargeClusterTest.test_consumer_offset_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64878#01965f1b-7772-47c1-b042-1bbcb0d980ee | FLAKY | 20/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64878#01965f09-4fc2-4076-8666-235f5636390d | FLAKY | 12/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64878#01965f09-4fc2-40cb-aa66-36c6b89fdec6 | FLAKY | 16/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65505">build#65505</a></summary>


| test_id                                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65505#01969414-917e-435d-9b33-ef5acf891027 | FLAKY         | 18/21    |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.S3           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65505#01969417-2c02-4909-9e64-ec7f58c1800e | FLAKY         | 5/21     |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65505#01969414-9180-4150-821e-162dda7a4b9d | FLAKY         | 20/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65505#01969414-9180-4150-821e-162dda7a4b9d | FLAKY         | 14/21    |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65505#01969417-2c01-4c41-8978-69950385941e | FLAKY         | 11/21    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65648">build#65648</a></summary>


| test_id                                                                                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves.cleanup_policy=compact.delete                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65648#0196ac1d-f13b-469d-a218-20699ca2191e | FLAKY         | 18/21    |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65648#0196ac1d-f139-4aed-b373-1494ae089b99 | FLAKY         | 13/21    |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.ABS                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65648#0196ac40-1e77-4b6e-8722-afa388084054 | FLAKY         | 2/21     |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.S3                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65648#0196ac40-1e79-453b-afaf-b58c80707564 | FLAKY         | 2/21     |
| rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type_and_url_style=.CloudStorageType.ABS.2.virtual_host.test_case=.TS_Read==True.TS_TxRangeMaterialized==True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65648#0196ac40-1e79-4ff2-8845-1424f3e4c1b5 | FAIL          | 0/1      |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65648#0196ac40-1e77-4b6e-8722-afa388084054 | FLAKY         | 15/21    |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65710">build#65710</a></summary>


| test_class                       | test_method   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                       |
|:---------------------------------|:--------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowIndexingCompactedTopicTest | test_upload   | {"cloud_storage_type": 2} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65710#0196b119-3e44-417a-a4ff-a2d4260f46a3 | FLAKY         | 12/70    | upstream reliability is '100.0'. current run reliability is '6.451612903225806'. drift is 93.54839 and the allowed drift is set to 50. The test should FAIL  |
| ShadowIndexingCompactedTopicTest | test_upload   | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65710#0196b119-3e45-4020-8bdc-aba625679e5a | FLAKY         | 14/70    | upstream reliability is '100.0'. current run reliability is '3.4482758620689653'. drift is 96.55172 and the allowed drift is set to 50. The test should FAIL |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65839">build#65839</a></summary>


| test_class                       | test_method                     | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:---------------------------------|:--------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataTransformsTest               | test_consume_from_offset        | {"offset": "+0"}                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65839#0196c5e5-bf3d-4296-b24f-debe819e26e1 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| PartitionBalancerTest            | test_unavailable_nodes          |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65839#0196c5ff-7621-4e6f-af60-3ba4bd2a1615 | FLAKY         | 14/21    | upstream reliability is '71.0217755443886'. current run reliability is '66.66666666666666'. drift is 4.35511 and the allowed drift is set to 50. The test should PASS   |
| RandomNodeOperationsTest         | test_node_operations            | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65839#0196c5ff-761e-4ff0-92df-6b2d5f6f9d0f | FLAKY         | 20/21    | upstream reliability is '97.17741935483872'. current run reliability is '95.23809523809523'. drift is 1.93932 and the allowed drift is set to 50. The test should PASS  |
| ScalingUpTest                    | test_adding_nodes_to_cluster    | {"partition_count": 20}                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65839#0196c5ff-761e-4ff0-92df-6b2d5f6f9d0f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| ShadowIndexingCompactedTopicTest | test_upload                     | {"cloud_storage_type": 2}                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65839#0196c5e5-bf3e-43a1-8fd0-48b8831c47ea | FLAKY         | 2/21     | upstream reliability is '100.0'. current run reliability is '9.523809523809524'. drift is 90.47619 and the allowed drift is set to 50. The test should FAIL             |
| ShadowIndexingCompactedTopicTest | test_upload                     | {"cloud_storage_type": 1}                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65839#0196c5e5-bf3c-4c15-846e-1074f7f5fe75 | FLAKY         | 5/21     | upstream reliability is '100.0'. current run reliability is '23.809523809523807'. drift is 76.19048 and the allowed drift is set to 50. The test should FAIL            |
| UpgradeBackToBackTest            | test_upgrade_with_all_workloads | {"single_upgrade": false}                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65839#0196c5e5-bf3e-43a1-8fd0-48b8831c47ea | FLAKY         | 14/21    | upstream reliability is '86.58649398704902'. current run reliability is '66.66666666666666'. drift is 19.91983 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65974">build#65974</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65974#0196cde9-11d3-473f-94e8-c3353472e005 | FLAKY         | 20/21    | upstream reliability is '95.53429027113238'. current run reliability is '95.23809523809523'. drift is 0.2962 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

Lazin commented on src/v/cluster/archival/segment_reupload.h:
```diff
@@ -47,26 +78,36 @@ class segment_collector {
       size_t max_uploaded_segment_size,
-      std::optional<model::offset> end_inclusive = std::nullopt);
+      std::optional<model::offset> end_inclusive,
+      std::optional<model::offset> end_exclusive,
```
if the 'end_exclusive' is an LSO does it have to be an optional?

oleiman replied:
> does it have to be an optional

i wouldn't think so. i made it optional because the reupload paths don't seem to need it - or put another way, to avoid updating other constructor call sites. 

i don't have a strong opinion about it though, beyond a vague sense of keeping the diff small.

Lazin reviewed (dismissed):
LGTM for the most part. The CI failure looks unrelated

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65505-->
<!--0a7e5372-e715-45eb-9f54-24e111f02338-->
<!--title-->
## Retry command for [Build#65505](https://buildkite.com/redpanda/redpanda/builds/65505)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
```
<!--/body-->

Lazin commented:
I think it fails because of this:
```
WARN  2025-05-03 03:58:41,846 [shard 1:au  ] archival - segment_reupload.cc:145 - Start offset 465970 is ahead of manifest last offset 465969 for ntp {kafka/panda-topic/0}, not a reupload
DEBUG 2025-05-03 03:58:41,846 [shard 1:au  ] archival - [fiber4 kafka/panda-topic/0] - ntp_archiver_service.cc:1762 - failed to make upload candidate: candidate creation error: candidate cannot replace manifest entry
```
so it looks like it tries to reupload the segment but includes more data than needed so the reuploaded segment doesn't line up with the one in the manifest

oleiman commented:
> I think it fails because of this:
> 
> ```
> WARN  2025-05-03 03:58:41,846 [shard 1:au  ] archival - segment_reupload.cc:145 - Start offset 465970 is ahead of manifest last offset 465969 for ntp {kafka/panda-topic/0}, not a reupload
> DEBUG 2025-05-03 03:58:41,846 [shard 1:au  ] archival - [fiber4 kafka/panda-topic/0] - ntp_archiver_service.cc:1762 - failed to make upload candidate: candidate creation error: candidate cannot replace manifest entry
> ```
> 
> so it looks like it tries to reupload the segment but includes more data than needed so the reuploaded segment doesn't line up with the one in the manifest

Ah, I thought this was already existing flake...will look into it

oleiman commented:
/ci-repeat 1
dt-repeat=50
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65648-->
<!--90aae0e8-17b7-415d-8a9e-b79e856840f9-->
<!--5eb8bdbe-6e14-4bb9-88f4-0ca104276101-->
<!--title-->
## Retry command for [Build#65648](https://buildkite.com/redpanda/redpanda/builds/65648)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":2}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
skip-redpanda-build
skip-units
dt-repeat=50
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":2}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65710-->
<!--3c86ce23-00e8-4685-8736-27d805a621fb-->
<!--title-->
## Retry command for [Build#65710](https://buildkite.com/redpanda/redpanda/builds/65710)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
debug
skip-units

oleiman commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65839-->
<!--be6b7ee7-3ec5-41c7-800e-99df2b11eb43-->
<!--b6f7a414-f33f-4f2c-ad79-a47393230cbf-->
<!--title-->
## Retry command for [Build#65839](https://buildkite.com/redpanda/redpanda/builds/65839)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":2}
```
<!--/body-->

oleiman committed:
archival: Rejigger segment collector concurrency control

For non-reupload segment collection, it is not uncommon to arrive in a state
where the tail of a collected segment list is an unsealed segment. This
generally means that time based upload was triggered, and we are effectively
forcing this segment up to s3.

This is desirable behavior in general, but it has an undesirable interaction
with concurrency control code in make_upload_candidate. Since the tail segment
is open, there is a greater chance that it's generation ID will be incremented
(e.g. by some append) between discovery and read-lock acquisition, leading to
the entire upload being discarded. This happens more frequently than, for
example, compaction kicking in on already collected segments and can lead to
lots of upload candidates being discarded.

This commit adds a special case to existing concurrency control code for
exactly this situation. That is, we skip the generation ID check if and only if

- the last segment in the list was NOT closed AND
- the last segment is the ONLY segment with gen ID difference

In this case, it should be safe to assume that compaction hasn't taken place
and that, once we have acquired read locks, the size of the collected segments
on disk will not have changed. Therefore we proceed with the upload.

In any other case where there is a generation ID mismatch, we short circuit
the upload as usual.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
archival/tests: Bump ntp_archiver_test timeout

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65949-->
<!--72db6483-b8f8-4f8d-9f07-63aa6bff1bc7-->
<!--title-->
## Retry command for [Build#65949](https://buildkite.com/redpanda/redpanda/builds/65949)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":2}
```
<!--/body-->

oleiman commented:
/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload

Lazin reviewed (approved):
LGTM
